pkg/providers/instanceprofile/instanceprofile.go

```diff
@@ -0,0 +1,159 @@
+	return &Provider{
+		region: region,
+		iamapi: iamapi,
+		ec2api: ec2api,
+		cache:  cache,
+	}
+}
+
+func (p *Provider) Create(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (string, error) {
+	tags := lo.Assign(nodeClass.Spec.Tags, map[string]string{
+		fmt.Sprintf("kubernetes.io/cluster/%s", settings.FromContext(ctx).ClusterName): "owned",
+		corev1beta1.ManagedByAnnotationKey:                                             settings.FromContext(ctx).ClusterName,
+		v1beta1.LabelNodeClass:                                                         nodeClass.Name, v1.LabelTopologyRegion: p.region,
+	})
+	profileName := GetProfileName(ctx, nodeClass)
+
+	// An instance profile exists for this NodeClass
+	if _, ok := p.cache.Get(string(nodeClass.UID)); ok {
+		return profileName, nil
+	}
+	// Validate if the instance profile exists and has the correct role assigned to it
+	var instanceProfile *iam.InstanceProfile
+	out, err := p.iamapi.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{InstanceProfileName: aws.String(profileName)})
+	if err != nil {
+		if !awserrors.IsNotFound(err) {
+			return "", fmt.Errorf("getting instance profile %q, %w", profileName, err)
+		}
+		o, err := p.iamapi.CreateInstanceProfileWithContext(ctx, &iam.CreateInstanceProfileInput{
+			InstanceProfileName: aws.String(profileName),
+			Tags:                lo.MapToSlice(tags, func(k, v string) *iam.Tag { return &iam.Tag{Key: aws.String(k), Value: aws.String(v)} }),
+		})
+		if err != nil {
+			return "", fmt.Errorf("creating instance profile %q, %w", profileName, err)
+		}
+		instanceProfile = o.InstanceProfile
+	} else {
+		instanceProfile = out.InstanceProfile
+	}
+	if len(instanceProfile.Roles) == 1 {
+		return profileName, nil
+	}
+	if _, err = p.iamapi.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
+		InstanceProfileName: aws.String(profileName),
+		RoleName:            aws.String(nodeClass.Spec.Role),
+	}); err != nil {
+		return "", fmt.Errorf("adding role %q to instance profile %q, %w", nodeClass.Spec.Role, profileName, err)
+	}
+	p.cache.SetDefault(string(nodeClass.UID), nil)
+	return profileName, nil
+}
+
+func (p *Provider) AssociatedInstances(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) ([]string, error) {
+	profileName := GetProfileName(ctx, nodeClass)
+
+	// Get all instances that are using our instance profile name and are not yet terminated
+	var ids []string
+	if err := p.ec2api.DescribeInstancesPagesWithContext(ctx, &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("iam-instance-profile.name"),
+				Values: aws.StringSlice([]string{profileName}),
+			},
+			instanceStateFilter,
+		},
+	}, func(page *ec2.DescribeInstancesOutput, _ bool) bool {
+		for _, res := range page.Reservations {
+			ids = append(ids, lo.Map(res.Instances, func(i *ec2.Instance, _ int) string {
+				return aws.StringValue(i.InstanceId)
+			})...)
+		}
```

Comment:
nit: can do lo.Map()

Reply:
Not a huge fan of it since it's a double `lo.Map` which means you also have to do a `lo.Flatten` at the end of it all